### ISSUE TITLE
cmd/exec: adding CLONE_NEWPID to Unshareflags breaks execs

### ISSUE THREAD

**Issue by rminnich** [golang/go#19716]
### What version of Go are you using (`go version`)?
any version up to and including Go 1.8. It's not necessarily a Go issue, it's the sometimes
non-intuitive behavior or NEWPID namespaces with unshare(2).

### What operating system and processor architecture are you using (`go env`)?
Ubuntu 16.06 LTS; LInux 4.4.0; amd64

### What did you do?

Here's a simple program, which when run as root on Linux, will fail in a surprising way. I've left out error checking for space reasons. You can't test this on the playground so I put it here.

package main

import (
	"os"
	"os/exec"
	"syscall"
)

func main() {
	c := exec.Command(os.Args[1], os.Args[2:]...)
	c.Stdin, c.Stdout, c.Stderr = os.Stdin, os.Stdout, os.Stderr
        c.SysProcAttr = &syscall.SysProcAttr{Unshareflags: syscall.CLONE_NEWPID}
	c.Run()
}

Here's what happens when you run it with itself as the first thing to run. I'm running it with itself as the first argument to show the issues with starting a Go program in a CLONE_NEWPID namespace. 
If you simply run /bin/bash as the first argument it works fine. But what if, as in u-root, we have an init process written in Go and we wish to test it via private pid namespaces?  We can't.

rminnich@akaros:~/tmp/unsharefail$ sudo ./unsharefail  ./unsharefail /bin/bash
runtime: failed to create new OS thread (have 2 already; errno=22)
fatal error: newosproc

runtime stack:
runtime.throw(0x4be7ae, 0x9)
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/panic.go:596 +0x95
runtime.newosproc(0xc420026000, 0xc420036000)
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/os_linux.go:163 +0x185
runtime.newm(0x4c4b70, 0x0)
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/proc.go:1614 +0x12d
runtime.main.func1()
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/proc.go:126 +0x36
runtime.systemstack(0x53bb00)
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/asm_amd64.s:332 +0x79
runtime.mstart()
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/proc.go:1115

goroutine 1 [running]:
runtime.systemstack_switch()
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/asm_amd64.s:286 fp=0xc420022788 sp=0xc420022780
runtime.main()
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/proc.go:125 +0x6c fp=0xc4200227e0 sp=0xc420022788
runtime.goexit()
	/home/rminnich/projects/u-root/go1.6dynamic/src/runtime/asm_amd64.s:2152 +0x1 fp=0xc4200227e8 sp=0xc4200227e0
rminnich@akaros:~/tmp/unsharefail$ 

Script done on Sat 25 Mar 2017 12:06:26 PM PDT

### What did you expect to see?

A bash prompt.

### What did you see instead?

A failure in the Go runtime. The reason is the rather strange rules of CLONE_NEWPID in unshare: the child only gets one fork, so you generally immediately need to do another one; subsequent ones return ENOMEM (true!). Since the Go runtime forks several times at startup, it fails, after the first fork. It would probably be best, when CLONE_NEWPID is set in Unshareflags, to instead move it to cloneflags.

A further problem is the behavior of proc. The intent of a NEWPID namespace is that the spawned process have PID 1. But it does not until /proc is remounted. The weirdness here is that mounts of /proc were always idempotent. But in a CLONE_NEWPID namespace, the *first* mount of /proc after unshare results in a proc with the private PIDs. 

More fun is that if you do the mount of /proc in the CLONE_NEWPID namespace, without having done a CLONE_NEWNS, you'll break /proc, and hence ps and many other things, outside the NEWPID namespace. 

You can try this yourself, but be sure you have another window with su running to see the effects or you'll need to reboot on some distros, as you can't even su when proc goes wrong:
window 1:
sudo unshare -p /bin/bash
/bin/bash # so you can run more than one more command
ps
mount -t proc proc /proc
ps

In the other window:
ps # note it doesn't work out well

in window 1: 
exit

in the other window, depending on your distro, there's no proc mounted any more. Fun!

Hence, we probably, when faced with CLONE_NEWPID in unshare flags, add CLONE_NEWNS to the flags, and while we're at it just go ahead and mount /proc for the child, so that the NEWPID namespace functions in the way it was intended to.

I'm working on a fix.

**Comment by jessfraz** [comment]
I made a patch to update the `Unshareflags` docs to define this behavior.

**Comment by agnivade** [comment]
@jessfraz - I see that the CL was abandoned with the reason as "unneeded". Should this issue be closed too ?

**Comment by andybons** [comment]
From @bradfitz on the CL:

> In that case, let's drop this. Not only is it time consuming (and an uphill battle) to mirror the Linux manpages into the Go docs, but it detracts from Go-specific notes we want to document.

Closing in favor of this.
